src/solvers/dg/3d/dg.jl

```diff
@@ -1307,11 +1319,120 @@ function rhs!(dg::Dg3D, t_stage)
+      u_node = get_node_vars(dg.elements.u, dg, i, j, k, element_id)
+      U1_mean += u_node[1] * weights[i] * weights[j] * weights[k] / 8.0
+      U2_mean += u_node[2] * weights[i] * weights[j] * weights[k] / 8.0 
+      U3_mean += u_node[3] * weights[i] * weights[j] * weights[k] / 8.0
+      U4_mean += u_node[4] * weights[i] * weights[j] * weights[k] / 8.0
+      U5_mean += u_node[5] * weights[i] * weights[j] * weights[k] / 8.0
+    end
+    # Detect if limiting is necessary
+    Theta_rho = 1
+    if rho_min < pp_limiter_threshold 
+      Theta_rho = (U1_mean - pp_limiter_threshold) / (U1_mean - rho_min)
+    end
+    Theta = min(1, Theta_rho)
+    # Apply limiter if necessary
+    if Theta < 1
+      #println("apply PP limiter 1 ", Theta_rho)
+      for k in 1:nnodes(dg), j in 1:nnodes(dg), i in 1:nnodes(dg)
+        u_node = get_node_vars(dg.elements.u, dg, i, j, k, element_id)
+        dg.elements.u[1,i,j,k,element_id] = Theta * u_node[1] + (1-Theta) * U1_mean 
+        dg.elements.u[2,i,j,k,element_id] = Theta * u_node[2] + (1-Theta) * U2_mean 
+        dg.elements.u[3,i,j,k,element_id] = Theta * u_node[3] + (1-Theta) * U3_mean 
+        dg.elements.u[4,i,j,k,element_id] = Theta * u_node[4] + (1-Theta) * U4_mean 
+        dg.elements.u[5,i,j,k,element_id] = Theta * u_node[5] + (1-Theta) * U5_mean 
+      end
+    end
+  end
+end
+
+# Apply positivity limiter of Zhan and Shu to nodal values elements.u
+# pressure
+function apply_positivity_preserving_limiter_pressure!(u, dg::Dg3D, equation::CompressibleEulerEquations3D)
+  @unpack weights, pp_limiter_threshold = dg
+  gamma = equation.gamma
+
+  Threads.@threads for element_id in 1:dg.n_elements
+    # Dermine minimum density, pressure and maximum convex entropy estimate
+    p_min = 0.0
+    for k in 1:nnodes(dg), j in 1:nnodes(dg), i in 1:nnodes(dg)
+      u_node = get_node_vars(dg.elements.u, dg, i, j, k, element_id)
+      p = (gamma-1) * (u_node[5] - 0.5 * (u_node[2]^2 + u_node[3]^2 + u_node[4]^2) / u_node[1])
+      p_min = min(p_min, p)
+    end
+    # Determine mean value of element
+    U1_mean = 0.0
+    U2_mean = 0.0
+    U3_mean = 0.0
+    U4_mean = 0.0
+    U5_mean = 0.0
+    for k in 1:nnodes(dg), j in 1:nnodes(dg), i in 1:nnodes(dg)
+      u_node = get_node_vars(dg.elements.u, dg, i, j, k, element_id)
+      U1_mean += u_node[1] * weights[i] * weights[j] * weights[k] / 8.0
+      U2_mean += u_node[2] * weights[i] * weights[j] * weights[k] / 8.0 
+      U3_mean += u_node[3] * weights[i] * weights[j] * weights[k] / 8.0
+      U4_mean += u_node[4] * weights[i] * weights[j] * weights[k] / 8.0
+      U5_mean += u_node[5] * weights[i] * weights[j] * weights[k] / 8.0
+    end
+    # Detect if limiting is necessary
+    p_mean = (gamma-1) * (U5_mean - 0.5 * (U2_mean^2 + U3_mean^2 + U4_mean^2) / U1_mean)
+    Theta_p = 1
+    if p_min < pp_limiter_threshold 
+      Theta_p = (p_mean - pp_limiter_threshold) / (p_mean - p_min)
```

Comment:
Does this really result in a sharp estimate? Or could some additional logic help?

Reply:
No, the estimate for the pressure is based on Jensen's inequality...

Reply:
Or maybe I do not understand your question?

Reply:
No, I think it's fine.